cirq/circuits/circuit.py

```diff
@@ -763,12 +758,41 @@ def are_all_matches_terminal(self, predicate: Callable[['cirq.Operation'], bool]
 
         Returns:
             Whether or not all `Operation` s in a circuit that satisfy the
-            given predicate are terminal.
+            given predicate are terminal. Also checks within any CircuitGates
+            the circuit may contain.
         """
-        return all(
+        from cirq.circuits import CircuitOperation
+
+        # TaggedOperations can wrap CircuitOperations.
+        def is_circuit_op(op):
+            if isinstance(op, CircuitOperation):
+                return True
+            if isinstance(op, ops.TaggedOperation):
+                return is_circuit_op(op.sub_operation)
+            return False
```

Comment:
Also, I should say that we should document here that we only apply the given predicate to operations that are not `CircuitOperations` and that we instead transparently recurse into `CircuitOperations` looking for matches. If the use tried to call `circuit.are_all_matches_terminal(lambda op: isinstance(op, cirq.CircuitOperation))` that wouldn't work as expected with with implementation (it would never match anything).

Reply:
Done and done. Added a test for that `are_all_matches_terminal` case - seems like documenting the behavior is a reasonable "solution" for now.